### ISSUE TITLE
CMS 5: POST request to /admin/graphql redirected with trailing slash in live mode

### ISSUE THREAD

**Issue by chrispenny** [silverstripe/silverstripe-framework#10775]
## Affected Version

CMS 5 stable

## Setup (simplified)

https://github.com/silverstripe/silverstripe-linkfield

Extension:
```php
class SiteConfigExtension extends DataExtension
{
    private static array $has_one = [
        'CtaLink' => Link::class,
    ];

    public function updateCMSFields(FieldList $fields): void
    {
        $fields->addFieldToTab(
            'Root.Footer',
            LinkField::create('CtaLink')
        );
    }
}
```

Apply extension:
```yaml
SilverStripe\SiteConfig\SiteConfig:
  extensions:
    - App\Extensions\SiteConfigExtension
```

Trailing slashes enabled (not sure if this is relevant for `admin`):
```yml
SilverStripe\Control\Controller:
  add_trailing_slash: true
```

Environment type:
```
SS_ENVIRONMENT_TYPE="live"
```

## Description

* Open browser Network tab, limit to XHR requests.
* Access the Settings area of the CMS and then switch to the Tab containing the `LinkField`.
* See 2 requests:
   * One `POST` request containing our payload to `https://domain/admin/graphql` that gets a `301` response.
   * One `GET` request **now missing the payload** to `https://domain/admin/graphql/` (i.e. with a trailing slash) that now gets a `200` response.

**Result:**
```
{"errors":[{"message":"This endpoint requires a \"query\" parameter"}]}
```

## Debugging

I noticed that the `CanonicalURLMiddleware` is only applied on live environments by default:
```php
protected $enabledEnvs = [
    CoreKernel::LIVE
];
```

I updated one of our other (working) environments to `SS_ENVIRONMENT_TYPE="live"`, and then started to experience the issue on that environment as well.

I then tried removing the `add_trailing_slash` config, and also specifically disabled it in `CanonicalURLMiddleware`:

```yaml
SilverStripe\Core\Injector\Injector:
  SilverStripe\Control\Middleware\CanonicalURLMiddleware:
    properties:
      enforceTrailingSlashConfig: false
```

I deployed this to a `live` environment, and the redirect no longer occurs.

## PRs
- https://github.com/silverstripe/silverstripe-framework/pull/10781

**Comment by GuySartorelli** [comment]
Additional context: The above mentioned `LinkField` - but this is not limited to `LinkField`. This also can be seen by simply visiting the `/admin/assets` admin from `silverstripe/asset-admin`.

**Comment by GuySartorelli** [comment]
This is technically a `impact/medium` according to https://docs.silverstripe.org/en/5/contributing/code/#labels because there are workarounds available (disable the redirect or don't add trailing slashes in the first place) - but it's also pretty egregious regardless of label.

**Comment by chrispenny** [comment]
> disable the redirect or don't add trailing slashes in the first place

Respectfully, I would argue these aren't workarounds for projects that need to use the trailing slash feature.

Is there a config change / workaround that doesn't also affect frontend behaviour?

**Comment by GuySartorelli** [comment]
That's what the `enforceTrailingSlashConfigIgnorePaths` property is meant to do, which is what seems to be broken. I take your point that if you _need_ the trailing slash and you _need_ the redirects on the front-end, then there's not really any valid workaround. I'll bump to impact high with that context.

**Comment by GuySartorelli** [comment]
@chrispenny This has been fixed and tagged as [5.0.1](https://github.com/silverstripe/silverstripe-framework/releases/tag/5.0.1)
